File: frontend/mobile/src/services/ConfigurationService.js
/**
 * WildCAM Mobile - Configuration Service
 * Handles remote configuration and local settings
 * 
 * @author WildCAM ESP32 Team
 * @version 3.0.0
 * @date 2025
 */

import AsyncStorage from '@react-native-async-storage/async-storage';
import APIService from './APIService';

/**
 * Configuration keys for local storage
 */
const CONFIG_KEYS = {
  API_BASE_URL: 'config_api_base_url',
  WEBSOCKET_URL: 'config_websocket_url',
  CAPTURE_INTERVAL: 'config_capture_interval',
  MOTION_SENSITIVITY: 'config_motion_sensitivity',
  ENABLE_NIGHT_MODE: 'config_enable_night_mode',
  ENABLE_CLOUD_UPLOAD: 'config_enable_cloud_upload',
  JPEG_QUALITY: 'config_jpeg_quality',
  FRAME_SIZE: 'config_frame_size',
  REMOTE_CONFIG: 'config_remote',
  REMOTE_CONFIG_TIME: 'config_remote_time',
  SETTINGS_VERSION: 'config_settings_version',
};

/**
 * Default configuration values
 */
const DEFAULT_CONFIG = {
  apiBaseURL: 'https://api.wildlife-monitor.com',
  websocketURL: 'wss://api.wildlife-monitor.com/ws',
  captureInterval: 300,
  motionSensitivity: 50,
  enableNightMode: true,
  enableCloudUpload: true,
  jpegQuality: 10,
  frameSize: 'UXGA',
};

/**
 * Configuration presets
 */
const PRESETS = {
  highQualityResearch: {
    name: 'High Quality Research',
    description: 'Best quality images, shorter intervals',
    captureInterval: 300,
    motionSensitivity: 70,
    enableNightMode: true,
    jpegQuality: 8,
    frameSize: 'UXGA',
  },
  longTermMonitoring: {
    name: 'Long-term Monitoring',
    description: 'Balanced quality and battery life',
    captureInterval: 600,
    motionSensitivity: 50,
    enableNightMode: true,
    jpegQuality: 15,
    frameSize: 'SXGA',
  },
  maximumBatteryLife: {
    name: 'Maximum Battery Life',
    description: 'Extended battery operation',
    captureInterval: 1800,
    motionSensitivity: 40,
    enableNightMode: false,
    jpegQuality: 20,
    frameSize: 'SVGA',
  },
  quickSurvey: {
    name: 'Quick Survey',
    description: 'Fast capture, moderate quality',
    captureInterval: 60,
    motionSensitivity: 80,
    enableNightMode: true,
    jpegQuality: 12,
    frameSize: 'XGA',
  },
};

/**
 * Frame size options with resolution info
 */
const FRAME_SIZES = {
  UXGA: { name: 'UXGA', resolution: '1600x1200', description: 'Highest quality' },
  SXGA: { name: 'SXGA', resolution: '1280x1024', description: 'High quality' },
  XGA: { name: 'XGA', resolution: '1024x768', description: 'Good quality' },
  SVGA: { name: 'SVGA', resolution: '800x600', description: 'Standard quality' },
  VGA: { name: 'VGA', resolution: '640x480', description: 'Basic quality' },
};

/**
 * ConfigurationService class for managing app and device configuration
 */
class ConfigurationService {
  constructor() {
    this.config = { ...DEFAULT_CONFIG };
    this.remoteConfig = null;
    this.lastRemoteConfigTime = null;
    this.configCacheDuration = 3600000; // 1 hour in milliseconds
    this.listeners = [];
  }

  /**
   * Initialize the configuration service
   */
  async initialize() {
    await this.loadLocalConfig();
    await this.loadCachedRemoteConfig();
    
    // Try to fetch remote config (non-blocking)
    this.fetchRemoteConfig().catch(console.error);
  }

  /**
   * Load local configuration from storage
   */
  async loadLocalConfig() {
    try {
      const keys = Object.values(CONFIG_KEYS);
      const results = await AsyncStorage.multiGet(keys);
      
      const configMap = {};
      results.forEach(([key, value]) => {
        if (value !== null) {
          try {
            configMap[key] = JSON.parse(value);
          } catch {
            configMap[key] = value;
          }
        }
      });

      // Apply loaded values
      if (configMap[CONFIG_KEYS.API_BASE_URL]) {
        this.config.apiBaseURL = configMap[CONFIG_KEYS.API_BASE_URL];
        APIService.setBaseURL(this.config.apiBaseURL);
      }
      if (configMap[CONFIG_KEYS.WEBSOCKET_URL]) {
        this.config.websocketURL = configMap[CONFIG_KEYS.WEBSOCKET_URL];
      }
      if (configMap[CONFIG_KEYS.CAPTURE_INTERVAL] !== undefined) {
        this.config.captureInterval = configMap[CONFIG_KEYS.CAPTURE_INTERVAL];
      }
      if (configMap[CONFIG_KEYS.MOTION_SENSITIVITY] !== undefined) {
        this.config.motionSensitivity = configMap[CONFIG_KEYS.MOTION_SENSITIVITY];
      }
      if (configMap[CONFIG_KEYS.ENABLE_NIGHT_MODE] !== undefined) {
        this.config.enableNightMode = configMap[CONFIG_KEYS.ENABLE_NIGHT_MODE];
      }
      if (configMap[CONFIG_KEYS.ENABLE_CLOUD_UPLOAD] !== undefined) {
        this.config.enableCloudUpload = configMap[CONFIG_KEYS.ENABLE_CLOUD_UPLOAD];
      }
      if (configMap[CONFIG_KEYS.JPEG_QUALITY] !== undefined) {
        this.config.jpegQuality = configMap[CONFIG_KEYS.JPEG_QUALITY];
      }
      if (configMap[CONFIG_KEYS.FRAME_SIZE]) {
        this.config.frameSize = configMap[CONFIG_KEYS.FRAME_SIZE];
      }
    } catch (error) {
      console.error('Failed to load local config:', error);
    }
  }

  /**
   * Load cached remote configuration
   */
  async loadCachedRemoteConfig() {
    try {
      const cachedConfig = await AsyncStorage.getItem(CONFIG_KEYS.REMOTE_CONFIG);
      const cachedTime = await AsyncStorage.getItem(CONFIG_KEYS.REMOTE_CONFIG_TIME);
      
      if (cachedConfig && cachedTime) {
        this.remoteConfig = JSON.parse(cachedConfig);
        this.lastRemoteConfigTime = new Date(JSON.parse(cachedTime));
      }
    } catch (error) {
      console.error('Failed to load cached remote config:', error);
    }
  }

  /**
   * Fetch remote configuration from server
   * @returns {Promise<Object>} - Remote configuration
   */
  async fetchRemoteConfig() {
    // Check if cache is still valid
    if (this.remoteConfig && this.lastRemoteConfigTime) {
      const elapsed = Date.now() - this.lastRemoteConfigTime.getTime();
      if (elapsed < this.configCacheDuration) {
        return this.remoteConfig;
      }
    }

    try {
      const response = await APIService.getRemoteConfig();
      this.remoteConfig = response;
      this.lastRemoteConfigTime = new Date();

      // Cache the remote config
      await AsyncStorage.setItem(CONFIG_KEYS.REMOTE_CONFIG, JSON.stringify(response));
      await AsyncStorage.setItem(CONFIG_KEYS.REMOTE_CONFIG_TIME, JSON.stringify(new Date().toISOString()));

      // Apply remote config to local settings
      this.applyRemoteConfig(response);

      // Notify listeners
      this.notifyListeners();

      return response;
    } catch (error) {
      console.error('Failed to fetch remote config:', error);
      throw error;
    }
  }

  /**
   * Apply remote configuration to local settings
   * @param {Object} remoteConfig - Remote configuration object
   */
  applyRemoteConfig(remoteConfig) {
    if (remoteConfig.capture_settings) {
      const { capture_settings } = remoteConfig;
      if (capture_settings.capture_delay !== undefined) {
        this.setConfigValue('captureInterval', capture_settings.capture_delay);
      }
      if (capture_settings.jpeg_quality !== undefined) {
        this.setConfigValue('jpegQuality', capture_settings.jpeg_quality);
      }
      if (capture_settings.frame_size !== undefined) {
        this.setConfigValue('frameSize', capture_settings.frame_size);
      }
      if (capture_settings.enable_night_mode !== undefined) {
        this.setConfigValue('enableNightMode', capture_settings.enable_night_mode);
      }
    }

    if (remoteConfig.upload_settings) {
      const { upload_settings } = remoteConfig;
      if (upload_settings.enable_auto_upload !== undefined) {
        this.setConfigValue('enableCloudUpload', upload_settings.enable_auto_upload);
      }
    }
  }

  // ==================== Getters ====================

  /**
   * Get all configuration values
   * @returns {Object} - Configuration object
   */
  getConfig() {
    return { ...this.config };
  }

  /**
   * Get a specific configuration value
   * @param {string} key - Configuration key
   * @returns {*} - Configuration value
   */
  getValue(key) {
    return this.config[key];
  }

  /**
   * Get API base URL
   * @returns {string} - API base URL
   */
  getAPIBaseURL() {
    return this.config.apiBaseURL;
  }

  /**
   * Get WebSocket URL
   * @returns {string} - WebSocket URL
   */
  getWebSocketURL() {
    return this.config.websocketURL;
  }

  /**
   * Get remote configuration
   * @returns {Object|null} - Remote configuration
   */
  getRemoteConfig() {
    return this.remoteConfig;
  }

  // ==================== Setters ====================

  /**
   * Set a configuration value
   * @param {string} key - Configuration key
   * @param {*} value - Configuration value
   */
  async setConfigValue(key, value) {
    this.config[key] = value;
    
    // Map to storage key and persist
    const storageKeyMap = {
      apiBaseURL: CONFIG_KEYS.API_BASE_URL,
      websocketURL: CONFIG_KEYS.WEBSOCKET_URL,
      captureInterval: CONFIG_KEYS.CAPTURE_INTERVAL,
      motionSensitivity: CONFIG_KEYS.MOTION_SENSITIVITY,
      enableNightMode: CONFIG_KEYS.ENABLE_NIGHT_MODE,
      enableCloudUpload: CONFIG_KEYS.ENABLE_CLOUD_UPLOAD,
      jpegQuality: CONFIG_KEYS.JPEG_QUALITY,
      frameSize: CONFIG_KEYS.FRAME_SIZE,
    };

    const storageKey = storageKeyMap[key];
    if (storageKey) {
      await AsyncStorage.setItem(storageKey, JSON.stringify(value));
    }

    // Special handling for API URL
    if (key === 'apiBaseURL') {
      APIService.setBaseURL(value);
    }

    // Notify listeners
    this.notifyListeners();
  }

  /**
   * Set multiple configuration values
   * @param {Object} values - Configuration values object
   */
  async setMultipleValues(values) {
    for (const [key, value] of Object.entries(values)) {
      await this.setConfigValue(key, value);
    }
  }

  // ==================== Device Configuration ====================

  /**
   * Push configuration to a specific device
   * @param {string} deviceId - Device ID
   * @returns {Promise<boolean>} - Success status
   */
  async pushConfigToDevice(deviceId) {
    try {
      const deviceConfig = {
        capture_interval: this.config.captureInterval,
        motion_sensitivity: this.config.motionSensitivity,
        enable_night_mode: this.config.enableNightMode,
        enable_cloud_upload: this.config.enableCloudUpload,
        jpeg_quality: this.config.jpegQuality,
        frame_size: this.config.frameSize,
      };

      await APIService.updateDeviceConfig(deviceId, deviceConfig);
      return true;
    } catch (error) {
      console.error('Failed to push config to device:', error);
      throw error;
    }
  }

  /**
   * Push configuration to multiple devices
   * @param {Array<string>} deviceIds - Array of device IDs
   * @returns {Promise<Object>} - Results for each device
   */
  async pushConfigToMultipleDevices(deviceIds) {
    const results = {};
    for (const deviceId of deviceIds) {
      try {
        await this.pushConfigToDevice(deviceId);
        results[deviceId] = { success: true };
      } catch (error) {
        results[deviceId] = { success: false, error: error.message };
      }
    }
    return results;
  }

  // ==================== Presets ====================

  /**
   * Get available presets
   * @returns {Object} - Available presets
   */
  getPresets() {
    return PRESETS;
  }

  /**
   * Apply a configuration preset
   * @param {string} presetKey - Preset key
   */
  async applyPreset(presetKey) {
    const preset = PRESETS[presetKey];
    if (!preset) {
      throw new Error(`Unknown preset: ${presetKey}`);
    }

    await this.setMultipleValues({
      captureInterval: preset.captureInterval,
      motionSensitivity: preset.motionSensitivity,
      enableNightMode: preset.enableNightMode,
      jpegQuality: preset.jpegQuality,
      frameSize: preset.frameSize,
    });
  }

  // ==================== Validation ====================

  /**
   * Validate configuration values
   * @returns {Array<string>} - Array of validation errors
   */
  validateConfig() {
    const errors = [];

    if (this.config.captureInterval < 1 || this.config.captureInterval > 3600) {
      errors.push('Capture interval must be between 1 and 3600 seconds');
    }

    if (this.config.motionSensitivity < 0 || this.config.motionSensitivity > 100) {
      errors.push('Motion sensitivity must be between 0 and 100');
    }

    if (this.config.jpegQuality < 1 || this.config.jpegQuality > 63) {
      errors.push('JPEG quality must be between 1 and 63');
    }

    if (!Object.keys(FRAME_SIZES).includes(this.config.frameSize)) {
      errors.push('Invalid frame size');
    }

    return errors;
  }

  // ==================== Export/Import ====================

  /**
   * Export configuration to JSON
   * @returns {string} - JSON configuration string
   */
  exportConfig() {
    return JSON.stringify({
      ...this.config,
      exportDate: new Date().toISOString(),
      version: '1.0',
    }, null, 2);
  }

  /**
   * Import configuration from JSON
   * @param {string} jsonString - JSON configuration string
   */
  async importConfig(jsonString) {
    try {
      const imported = JSON.parse(jsonString);
      
      // Validate and apply imported values
      const validKeys = ['captureInterval', 'motionSensitivity', 'enableNightMode', 
                         'enableCloudUpload', 'jpegQuality', 'frameSize'];
      
      for (const key of validKeys) {
        if (imported[key] !== undefined) {
          await this.setConfigValue(key, imported[key]);
        }
      }
    } catch (error) {
      throw new Error('Invalid configuration format');
    }
  }

  // ==================== Listeners ====================

  /**
   * Add a configuration change listener
   * @param {Function} listener - Listener function
   */
  addListener(listener) {
    this.listeners.push(listener);
  }

  /**
   * Remove a configuration change listener
   * @param {Function} listener - Listener function
   */
  removeListener(listener) {
    const index = this.listeners.indexOf(listener);
    if (index > -1) {
      this.listeners.splice(index, 1);
    }
  }

  /**
   * Notify all listeners of configuration change
   */
  notifyListeners() {
    this.listeners.forEach(listener => listener(this.config));
  }

  // ==================== Reset ====================

  /**
   * Reset configuration to defaults
   */
  async resetToDefaults() {
    this.config = { ...DEFAULT_CONFIG };
    
    const keys = Object.values(CONFIG_KEYS);
    await AsyncStorage.multiRemove(keys);
    
    this.notifyListeners();
  }
}

// Export singleton instance and utilities
const configService = new ConfigurationService();
export default configService;
export { CONFIG_KEYS, DEFAULT_CONFIG, PRESETS, FRAME_SIZES };

File: mobile/ios/Services/ConfigurationService.swift
/**
 * WildCAM iOS - Configuration Service
 * Handles remote configuration and local settings
 * 
 * @author WildCAM ESP32 Team
 * @version 1.0.0
 * @date 2025
 */

import Foundation
import Combine

// MARK: - Configuration Service
class ConfigurationService: ObservableObject {
    static let shared = ConfigurationService()
    
    // API configuration
    @Published var apiBaseURL: String
    @Published var websocketURL: String
    
    // Remote configuration
    @Published var remoteConfig: RemoteConfiguration?
    @Published var lastConfigUpdate: Date?
    @Published var isConfigLoaded = false
    
    // Local settings
    @Published var captureInterval: Int {
        didSet {
            UserDefaults.standard.set(captureInterval, forKey: "captureInterval")
        }
    }
    @Published var motionSensitivity: Int {
        didSet {
            UserDefaults.standard.set(motionSensitivity, forKey: "motionSensitivity")
        }
    }
    @Published var enableNightMode: Bool {
        didSet {
            UserDefaults.standard.set(enableNightMode, forKey: "enableNightMode")
        }
    }
    @Published var enableCloudUpload: Bool {
        didSet {
            UserDefaults.standard.set(enableCloudUpload, forKey: "enableCloudUpload")
        }
    }
    @Published var jpegQuality: Int {
        didSet {
            UserDefaults.standard.set(jpegQuality, forKey: "jpegQuality")
        }
    }
    @Published var frameSize: String {
        didSet {
            UserDefaults.standard.set(frameSize, forKey: "frameSize")
        }
    }
    
    // Configuration cache
    private let configCacheKey = "cachedRemoteConfig"
    private let configCacheTimeKey = "cachedRemoteConfigTime"
    private let configCacheDuration: TimeInterval = 3600 // 1 hour
    
    private var cancellables = Set<AnyCancellable>()
    
    private init() {
        // Load API URLs from Info.plist or use defaults
        let bundle = Bundle.main
        self.apiBaseURL = bundle.object(forInfoDictionaryKey: "API_BASE_URL") as? String
            ?? "https://api.wildlife-monitor.com"
        self.websocketURL = bundle.object(forInfoDictionaryKey: "WEBSOCKET_URL") as? String
            ?? "wss://api.wildlife-monitor.com/ws"
        
        // Load local settings from UserDefaults
        captureInterval = UserDefaults.standard.integer(forKey: "captureInterval")
        motionSensitivity = UserDefaults.standard.integer(forKey: "motionSensitivity")
        enableNightMode = UserDefaults.standard.bool(forKey: "enableNightMode")
        enableCloudUpload = UserDefaults.standard.bool(forKey: "enableCloudUpload")
        jpegQuality = UserDefaults.standard.integer(forKey: "jpegQuality")
        frameSize = UserDefaults.standard.string(forKey: "frameSize") ?? "UXGA"
        
        // Set defaults if not configured
        if captureInterval == 0 { captureInterval = 300 }
        if motionSensitivity == 0 { motionSensitivity = 50 }
        if jpegQuality == 0 { jpegQuality = 10 }
        if !UserDefaults.standard.bool(forKey: "settingsInitialized") {
            enableCloudUpload = true
            UserDefaults.standard.set(true, forKey: "settingsInitialized")
        }
        
        // Load cached configuration
        loadCachedConfiguration()
    }
    
    // MARK: - Remote Configuration
    
    func fetchConfiguration() async {
        // Check if cache is still valid
        if let lastUpdate = lastConfigUpdate,
           Date().timeIntervalSince(lastUpdate) < configCacheDuration,
           remoteConfig != nil {
            return
        }
        
        do {
            let config = try await APIService.shared.getRemoteConfiguration()
            
            await MainActor.run {
                self.remoteConfig = config
                self.lastConfigUpdate = Date()
                self.isConfigLoaded = true
                self.applyRemoteConfiguration(config)
                self.cacheConfiguration(config)
            }
        } catch {
            print("Failed to fetch remote configuration: \(error.localizedDescription)")
            // Use cached configuration if available
        }
    }
    
    private func applyRemoteConfiguration(_ config: RemoteConfiguration) {
        // Apply capture settings
        if captureInterval != config.captureSettings.captureDelay {
            captureInterval = config.captureSettings.captureDelay
        }
        if jpegQuality != config.captureSettings.jpegQuality {
            jpegQuality = config.captureSettings.jpegQuality
        }
        if frameSize != config.captureSettings.frameSize {
            frameSize = config.captureSettings.frameSize
        }
        if enableNightMode != config.captureSettings.enableNightMode {
            enableNightMode = config.captureSettings.enableNightMode
        }
        
        // Apply upload settings
        if enableCloudUpload != config.uploadSettings.enableAutoUpload {
            enableCloudUpload = config.uploadSettings.enableAutoUpload
        }
    }
    
    private func cacheConfiguration(_ config: RemoteConfiguration) {
        if let data = try? JSONEncoder().encode(config) {
            UserDefaults.standard.set(data, forKey: configCacheKey)
            UserDefaults.standard.set(Date(), forKey: configCacheTimeKey)
        }
    }
    
    private func loadCachedConfiguration() {
        guard let data = UserDefaults.standard.data(forKey: configCacheKey),
              let config = try? JSONDecoder().decode(RemoteConfiguration.self, from: data) else {
            return
        }
        
        remoteConfig = config
        lastConfigUpdate = UserDefaults.standard.object(forKey: configCacheTimeKey) as? Date
        isConfigLoaded = true
    }
    
    // MARK: - Push Configuration to Device
    
    func pushConfigurationToDevice(deviceId: String) async throws {
        let configuration: [String: Any] = [
            "capture_interval": captureInterval,
            "motion_sensitivity": motionSensitivity,
            "enable_night_mode": enableNightMode,
            "enable_cloud_upload": enableCloudUpload,
            "jpeg_quality": jpegQuality,
            "frame_size": frameSize
        ]
        
        _ = try await APIService.shared.updateCameraConfiguration(
            deviceId: deviceId,
            configuration: configuration
        )
    }
    
    // MARK: - Configuration Presets
    
    struct ConfigurationPreset {
        let name: String
        let description: String
        let captureInterval: Int
        let motionSensitivity: Int
        let enableNightMode: Bool
        let jpegQuality: Int
        let frameSize: String
    }
    
    static let presets: [ConfigurationPreset] = [
        ConfigurationPreset(
            name: "High Quality Research",
            description: "Best quality images, shorter intervals",
            captureInterval: 300,
            motionSensitivity: 70,
            enableNightMode: true,
            jpegQuality: 8,
            frameSize: "UXGA"
        ),
        ConfigurationPreset(
            name: "Long-term Monitoring",
            description: "Balanced quality and battery life",
            captureInterval: 600,
            motionSensitivity: 50,
            enableNightMode: true,
            jpegQuality: 15,
            frameSize: "SXGA"
        ),
        ConfigurationPreset(
            name: "Maximum Battery Life",
            description: "Extended battery operation",
            captureInterval: 1800,
            motionSensitivity: 40,
            enableNightMode: false,
            jpegQuality: 20,
            frameSize: "SVGA"
        ),
        ConfigurationPreset(
            name: "Quick Survey",
            description: "Fast capture, moderate quality",
            captureInterval: 60,
            motionSensitivity: 80,
            enableNightMode: true,
            jpegQuality: 12,
            frameSize: "XGA"
        )
    ]
    
    func applyPreset(_ preset: ConfigurationPreset) {
        captureInterval = preset.captureInterval
        motionSensitivity = preset.motionSensitivity
        enableNightMode = preset.enableNightMode
        jpegQuality = preset.jpegQuality
        frameSize = preset.frameSize
    }
    
    // MARK: - Validation
    
    func validateConfiguration() -> [String] {
        var errors: [String] = []
        
        if captureInterval < 1 || captureInterval > 3600 {
            errors.append("Capture interval must be between 1 and 3600 seconds")
        }
        
        if motionSensitivity < 0 || motionSensitivity > 100 {
            errors.append("Motion sensitivity must be between 0 and 100")
        }
        
        if jpegQuality < 1 || jpegQuality > 63 {
            errors.append("JPEG quality must be between 1 and 63")
        }
        
        let validFrameSizes = ["UXGA", "SXGA", "XGA", "SVGA", "VGA"]
        if !validFrameSizes.contains(frameSize) {
            errors.append("Invalid frame size")
        }
        
        return errors
    }
    
    // MARK: - Export/Import
    
    func exportConfiguration() -> Data? {
        let config: [String: Any] = [
            "capture_interval": captureInterval,
            "motion_sensitivity": motionSensitivity,
            "enable_night_mode": enableNightMode,
            "enable_cloud_upload": enableCloudUpload,
            "jpeg_quality": jpegQuality,
            "frame_size": frameSize,
            "export_date": ISO8601DateFormatter().string(from: Date())
        ]
        
        return try? JSONSerialization.data(withJSONObject: config, options: .prettyPrinted)
    }
    
    func importConfiguration(from data: Data) throws {
        guard let config = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw ConfigurationError.invalidFormat
        }
        
        if let interval = config["capture_interval"] as? Int {
            captureInterval = interval
        }
        if let sensitivity = config["motion_sensitivity"] as? Int {
            motionSensitivity = sensitivity
        }
        if let nightMode = config["enable_night_mode"] as? Bool {
            enableNightMode = nightMode
        }
        if let cloudUpload = config["enable_cloud_upload"] as? Bool {
            enableCloudUpload = cloudUpload
        }
        if let quality = config["jpeg_quality"] as? Int {
            jpegQuality = quality
        }
        if let size = config["frame_size"] as? String {
            frameSize = size
        }
    }
    
    enum ConfigurationError: LocalizedError {
        case invalidFormat
        case validationFailed([String])
        
        var errorDescription: String? {
            switch self {
            case .invalidFormat:
                return "Invalid configuration format"
            case .validationFailed(let errors):
                return "Configuration validation failed: \(errors.joined(separator: ", "))"
            }
        }
    }
}

// MARK: - Frame Size Options
enum FrameSize: String, CaseIterable {
    case uxga = "UXGA"
    case sxga = "SXGA"
    case xga = "XGA"
    case svga = "SVGA"
    case vga = "VGA"
    
    var resolution: String {
        switch self {
        case .uxga: return "1600x1200"
        case .sxga: return "1280x1024"
        case .xga: return "1024x768"
        case .svga: return "800x600"
        case .vga: return "640x480"
        }
    }
    
    var displayName: String {
        return "\(rawValue) (\(resolution))"
    }
}
